Windows issues with Transitive Dependency Resolver

The following error is thrown when using the Transitive Dependency Helper on windows machines. I believe the error is failing to convert to a URI for support across Linux/windows OS but I could be wrong and input is appreciated!


```[21:30:39] [Server thread/ERROR]: Error occurred while enabling RDQ v4.1.5 (Is it up to date?)
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.itsrainingplex.libs.com.alessiodp.libby.transitive.TransitiveDependencyHelper.findTransitiveLibraries(TransitiveDependencyHelper.java:170) ~[RDQ-4.1.5.jar:?]
	at com.itsrainingplex.libs.com.alessiodp.libby.LibraryManager.resolveTransitiveLibraries(LibraryManager.java:655) ~[RDQ-4.1.5.jar:?]
	at com.itsrainingplex.rdq.Settings.RManager_Paper.loadLibrary(RManager_Paper.java:21) ~[RDQ-4.1.5.jar:?]
	at com.itsrainingplex.rdq.Settings.Depends.loadWebJars(Depends.java:147) ~[RDQ-4.1.5.jar:?]
	at com.itsrainingplex.rdq.Settings.Depends.loadDepends(Depends.java:74) ~[RDQ-4.1.5.jar:?]
	at com.itsrainingplex.rdq.RDQ.onEnable(RDQ.java:39) ~[RDQ-4.1.5.jar:?]
	at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:281) ~[paper-api-1.20.4-R0.1-SNAPSHOT.jar:?]
	at io.papermc.paper.plugin.manager.PaperPluginInstanceManager.enablePlugin(PaperPluginInstanceManager.java:190) ~[paper-1.20.4.jar:git-Paper-365]
	at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.enablePlugin(PaperPluginManagerImpl.java:104) ~[paper-1.20.4.jar:git-Paper-365]
	at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:507) ~[paper-api-1.20.4-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.craftbukkit.v1_20_R3.CraftServer.enablePlugin(CraftServer.java:637) ~[paper-1.20.4.jar:git-Paper-365]
	at org.bukkit.craftbukkit.v1_20_R3.CraftServer.enablePlugins(CraftServer.java:548) ~[paper-1.20.4.jar:git-Paper-365]
	at net.minecraft.server.MinecraftServer.loadWorld0(MinecraftServer.java:673) ~[paper-1.20.4.jar:git-Paper-365]
	at net.minecraft.server.MinecraftServer.loadLevel(MinecraftServer.java:433) ~[paper-1.20.4.jar:git-Paper-365]
	at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:308) ~[paper-1.20.4.jar:git-Paper-365]
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1132) ~[paper-1.20.4.jar:git-Paper-365]
	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:321) ~[paper-1.20.4.jar:git-Paper-365]
	at java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.reflect.InvocationTargetException
	at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at com.itsrainingplex.libs.com.alessiodp.libby.transitive.TransitiveDependencyHelper.findTransitiveLibraries(TransitiveDependencyHelper.java:135) ~[RDQ-4.1.5.jar:?]
	... 17 more
Caused by: java.nio.file.InvalidPathException: Illegal char <:> at index 90: C:\Users\XXX\Desktop\Server\plugins\RDQ\lib\org\webjars\npm\csstype\maven-metadata-https:\repo1.maven.org\maven2\resolver-status.properties
	at sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:204) ~[?:?]
	at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:175) ~[?:?]
	at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77) ~[?:?]
	at sun.nio.fs.WindowsPath.parse(WindowsPath.java:92) ~[?:?]
	at sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:231) ~[?:?]
	at java.io.File.toPath(File.java:2401) ~[?:?]
	at org.eclipse.aether.internal.impl.DefaultTrackingFileManager.read(DefaultTrackingFileManager.java:56) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.DefaultUpdateCheckManager.read(DefaultUpdateCheckManager.java:461) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.DefaultUpdateCheckManager.checkMetadata(DefaultUpdateCheckManager.java:261) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.DefaultMetadataResolver.resolve(DefaultMetadataResolver.java:303) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.DefaultMetadataResolver.resolveMetadata(DefaultMetadataResolver.java:180) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.apache.maven.repository.internal.DefaultVersionRangeResolver.getVersions(DefaultVersionRangeResolver.java:189) ~[maven-resolver-provider-3.9.6.jar:3.9.6]
	at org.apache.maven.repository.internal.DefaultVersionRangeResolver.resolveVersionRange(DefaultVersionRangeResolver.java:142) ~[maven-resolver-provider-3.9.6.jar:3.9.6]
	at org.eclipse.aether.internal.impl.collect.DependencyCollectorDelegate.cachedResolveRangeResult(DependencyCollectorDelegate.java:406) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.collect.df.DfDependencyCollector.processDependency(DfDependencyCollector.java:202) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.collect.df.DfDependencyCollector.processDependency(DfDependencyCollector.java:156) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.collect.df.DfDependencyCollector.process(DfDependencyCollector.java:138) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.collect.df.DfDependencyCollector.doRecurse(DfDependencyCollector.java:343) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.collect.df.DfDependencyCollector.processDependency(DfDependencyCollector.java:277) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.collect.df.DfDependencyCollector.processDependency(DfDependencyCollector.java:156) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.collect.df.DfDependencyCollector.process(DfDependencyCollector.java:138) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.collect.df.DfDependencyCollector.doCollectDependencies(DfDependencyCollector.java:108) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.collect.DependencyCollectorDelegate.collectDependencies(DependencyCollectorDelegate.java:222) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies(DefaultDependencyCollector.java:87) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:328) ~[maven-resolver-impl-1.9.18.jar:1.9.18]
	at com.itsrainingplex.libs.com.alessiodp.libby.transitive.TransitiveDependencyCollector.findTransitiveDependencies(TransitiveDependencyCollector.java:96) ~[?:?]
	at com.itsrainingplex.libs.com.alessiodp.libby.transitive.TransitiveDependencyCollector.findTransitiveDependencies(TransitiveDependencyCollector.java:115) ~[?:?]
	at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at com.itsrainingplex.libs.com.alessiodp.libby.transitive.TransitiveDependencyHelper.findTransitiveLibraries(TransitiveDependencyHelper.java:135) ~[RDQ-4.1.5.jar:?]
	... 17 more
[21:30:39] [Server thread/INFO]: [RDQ] Disabling RDQ v4.1.5```

--- a/core/src/main/java/com/alessiodp/libby/Library.java
+++ b/core/src/main/java/com/alessiodp/libby/Library.java
@@ -224,7 +224,7 @@ public class Library {
      * @return true if library has classifier, false otherwise
      */
     public boolean hasClassifier() {
-        return classifier != null;
+        return classifier != null && !classifier.isEmpty();
     }
 
     /**

--- a/core/src/main/java/com/alessiodp/libby/Library.java
+++ b/core/src/main/java/com/alessiodp/libby/Library.java
@@ -5,12 +5,13 @@ import com.alessiodp.libby.relocation.Relocation;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
-
 import java.util.Base64;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.LinkedList;
 
+import static com.alessiodp.libby.Util.craftPartialPath;
+import static com.alessiodp.libby.Util.craftPath;
 import static com.alessiodp.libby.Util.hexStringToByteArray;
 import static com.alessiodp.libby.Util.replaceWithDots;
 import static java.util.Objects.requireNonNull;
@@ -146,13 +147,8 @@ public class Library {
         this.checksum = checksum;
         this.relocations = relocations != null ? Collections.unmodifiableList(new LinkedList<>(relocations)) : Collections.emptyList();
 
-        this.partialPath = this.groupId.replace('.', '/') + '/' + this.artifactId + '/' + version + '/';
-        String path = this.partialPath + this.artifactId + '-' + version;
-        if (hasClassifier()) {
-            path += '-' + classifier;
-        }
-
-        this.path = path + ".jar";
+        this.partialPath = craftPartialPath(this.artifactId, this.groupId, version);
+        this.path = craftPath(this.partialPath, this.artifactId, this.version, this.classifier);
 
         this.repositories = repositories != null ? Collections.unmodifiableList(new LinkedList<>(repositories)) : Collections.emptyList();
         relocatedPath = hasRelocations() ? path + "-relocated-" + Math.abs(this.relocations.hashCode()) + ".jar" : null;

--- a/core/src/main/java/com/alessiodp/libby/LibraryManager.java
+++ b/core/src/main/java/com/alessiodp/libby/LibraryManager.java
@@ -424,11 +424,7 @@ public abstract class LibraryManager {
             version = version.substring(0, version.length() - "-SNAPSHOT".length());
         }
 
-        String url = library.getPartialPath() + library.getArtifactId() + '-' + version + '-' + timestamp + '-' + buildNumber;
-        if (library.hasClassifier()) {
-            url += '-' + library.getClassifier();
-        }
-        return url + ".jar";
+        return Util.craftPath(library.getPartialPath(), library.getArtifactId(), version + '-' + timestamp + '-' + buildNumber, library.getClassifier());
     }
 
     /**

--- a/core/src/main/java/com/alessiodp/libby/Util.java
+++ b/core/src/main/java/com/alessiodp/libby/Util.java
@@ -1,6 +1,7 @@
 package com.alessiodp.libby;
 
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * Libby's utility class.
@@ -28,7 +29,7 @@ public final class Util {
      * @param string The string to convert
      * @return The byte array
      */
-    public static byte[] hexStringToByteArray(String string) {
+    public static byte[] hexStringToByteArray(@NotNull String string) {
         int len = string.length();
         byte[] data = new byte[len / 2];
         for (int i = 0; i < len; i += 2) {
@@ -36,5 +37,38 @@ public final class Util {
                     + Character.digit(string.charAt(i+1), 16));
         }
         return data;
+    }
+
+    /**
+     * Constructs the partial path of a {@link Library} given its artifactId, groupId and version.
+     *
+     * @param artifactId The artifactId of the library.
+     * @param groupId The groupId of the library.
+     * @param version The version of the library.
+     * @return The partial path of the library.
+     * @see Library#getPartialPath()
+     */
+    @NotNull
+    public static String craftPartialPath(@NotNull String artifactId, @NotNull String groupId, @NotNull String version) {
+        return groupId.replace('.', '/') + '/' + artifactId + '/' + version + '/';
+    }
+
+    /**
+     * Constructs the path of a {@link Library} given its partialPath, artifactId, version and classifier.
+     *
+     * @param partialPath The partialPath of the library.
+     * @param artifactId The artifactId of the library.
+     * @param version The version of the library.
+     * @param classifier The classifier of the library. May be null.
+     * @return The path of the library.
+     * @see Library#getPath()
+     */
+    @NotNull
+    public static String craftPath(@NotNull String partialPath, @NotNull String artifactId, @NotNull String version, @Nullable String classifier) {
+        String path = partialPath + artifactId + '-' + version;
+        if (classifier != null && !classifier.isEmpty()) {
+            path += '-' + classifier;
+        }
+        return path + ".jar";
     }
 }

--- a/core/src/main/java/com/alessiodp/libby/transitive/TransitiveDependencyCollector.java
+++ b/core/src/main/java/com/alessiodp/libby/transitive/TransitiveDependencyCollector.java
@@ -25,6 +25,7 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.Properties;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
@@ -34,6 +35,13 @@ import java.util.stream.Stream;
  * @see <a href=https://github.com/apache/maven-resolver>maven-resolver</a>.
  */
 class TransitiveDependencyCollector {
+
+    /**
+     * Counter used to generate ids for repositories
+     *
+     * @see #newDefaultRepository(String)
+     */
+    private static final AtomicInteger ID_GENERATOR = new AtomicInteger();
 
     /**
      * Maven repository system
@@ -72,7 +80,7 @@ class TransitiveDependencyCollector {
      */
     @NotNull
     public static RemoteRepository newDefaultRepository(@NotNull String url) {
-        return new RemoteRepository.Builder(url, "default", url).build();
+        return new RemoteRepository.Builder("repo" + ID_GENERATOR.getAndIncrement(), "default", url).build();
     }
 
     /**

--- a/core/src/main/java/com/alessiodp/libby/transitive/TransitiveDependencyCollector.java
+++ b/core/src/main/java/com/alessiodp/libby/transitive/TransitiveDependencyCollector.java
@@ -8,6 +8,7 @@ import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.artifact.DefaultArtifact;
 import org.eclipse.aether.collection.CollectRequest;
 import org.eclipse.aether.graph.Dependency;
+import org.eclipse.aether.repository.ArtifactRepository;
 import org.eclipse.aether.repository.LocalRepository;
 import org.eclipse.aether.repository.RemoteRepository;
 import org.eclipse.aether.resolution.ArtifactResult;
@@ -18,12 +19,15 @@ import org.eclipse.aether.supplier.RepositorySystemSupplier;
 import org.eclipse.aether.util.artifact.JavaScopes;
 import org.eclipse.aether.util.filter.ScopeDependencyFilter;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 import java.nio.file.Path;
+import java.util.AbstractMap.SimpleEntry;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
+import java.util.Map.Entry;
 import java.util.Properties;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.stream.Collectors;
@@ -91,11 +95,11 @@ class TransitiveDependencyCollector {
      * @param version      Maven dependency version
      * @param classifier   Maven artifact classifier. May be null
      * @param repositories Maven repositories that would be used for dependency resolution
-     * @return Transitive dependencies, exception otherwise
+     * @return Transitive dependencies paired with their repository url, exception otherwise
      * @throws DependencyResolutionException thrown if dependency doesn't exist on provided repositories
      */
     @NotNull
-    public Collection<Artifact> findTransitiveDependencies(@NotNull String groupId, @NotNull String artifactId, @NotNull String version, @NotNull String classifier, @NotNull List<RemoteRepository> repositories) throws DependencyResolutionException {
+    public Collection<Entry<Artifact, @Nullable String>> findTransitiveDependencies(@NotNull String groupId, @NotNull String artifactId, @NotNull String version, @NotNull String classifier, @NotNull List<RemoteRepository> repositories) throws DependencyResolutionException {
         Artifact artifact = new DefaultArtifact(groupId, artifactId, classifier, "jar", version);
 
         CollectRequest collectRequest = new CollectRequest(new Dependency(artifact, JavaScopes.COMPILE), repositories);
@@ -103,7 +107,18 @@ class TransitiveDependencyCollector {
 
         DependencyResult dependencyResult = repositorySystem.resolveDependencies(repositorySystemSession, dependencyRequest);
 
-        return dependencyResult.getArtifactResults().stream().filter(ArtifactResult::isResolved).map(ArtifactResult::getArtifact).collect(Collectors.toList());
+        return dependencyResult.getArtifactResults()
+                .stream()
+                .filter(ArtifactResult::isResolved)
+                .map(artifactResult -> {
+                    ArtifactRepository repo = artifactResult.getRepository();
+                    String url = null;
+                    if (repo instanceof RemoteRepository) {
+                        url = ((RemoteRepository) repo).getUrl();
+                    }
+                    return new SimpleEntry<>(artifactResult.getArtifact(), url);
+                })
+                .collect(Collectors.toList());
     }
 
     /**
@@ -114,12 +129,12 @@ class TransitiveDependencyCollector {
      * @param version      Maven artifact version
      * @param classifier   Maven artifact classifier. May be null
      * @param repositories Maven repositories for transitive dependencies search
-     * @return Transitive dependencies, exception otherwise
+     * @return Transitive dependencies paired with their repository url, exception otherwise
      * @throws DependencyResolutionException thrown if dependency doesn't exist on provided repositories
      * @see #findTransitiveDependencies(String, String, String, String, List)
      */
     @NotNull
-    public Collection<Artifact> findTransitiveDependencies(@NotNull String groupId, @NotNull String artifactId, @NotNull String version, @NotNull String classifier, @NotNull Stream<String> repositories) throws DependencyResolutionException {
+    public Collection<Entry<Artifact, @Nullable String>> findTransitiveDependencies(@NotNull String groupId, @NotNull String artifactId, @NotNull String version, @NotNull String classifier, @NotNull Stream<String> repositories) throws DependencyResolutionException {
         return findTransitiveDependencies(groupId, artifactId, version, classifier, repositories.map(TransitiveDependencyCollector::newDefaultRepository).collect(Collectors.toList()));
     }
 

--- a/core/src/main/java/com/alessiodp/libby/transitive/TransitiveDependencyHelper.java
+++ b/core/src/main/java/com/alessiodp/libby/transitive/TransitiveDependencyHelper.java
@@ -46,7 +46,7 @@ public class TransitiveDependencyHelper {
     /**
      * Reflected getter methods of Artifact class
      */
-    private final Method artifactGetGroupIdMethod, artifactGetArtifactIdMethod, artifactGetVersionMethod, artifactGetClassifierMethod;
+    private final Method artifactGetGroupIdMethod, artifactGetArtifactIdMethod, artifactGetBaseVersionMethod, artifactGetClassifierMethod;
 
     /**
      * LibraryManager instance, used in {@link #findTransitiveLibraries(Library)}
@@ -93,8 +93,8 @@ public class TransitiveDependencyHelper {
             artifactGetGroupIdMethod = artifactClass.getMethod("getGroupId");
             // org.eclipse.aether.artifact.Artifact#getArtifactId()
             artifactGetArtifactIdMethod = artifactClass.getMethod("getArtifactId");
-            // org.eclipse.aether.artifact.Artifact#getVersion()
-            artifactGetVersionMethod = artifactClass.getMethod("getVersion");
+            // org.eclipse.aether.artifact.Artifact#getBaseVersion()
+            artifactGetBaseVersionMethod = artifactClass.getMethod("getBaseVersion");
             // org.eclipse.aether.artifact.Artifact#getClassifier()
             artifactGetClassifierMethod = artifactClass.getMethod("getClassifier");
         } catch (ReflectiveOperationException e) {
@@ -141,7 +141,7 @@ public class TransitiveDependencyHelper {
             for (Object artifact : artifacts) {
                 String groupId = (String) artifactGetGroupIdMethod.invoke(artifact);
                 String artifactId = (String) artifactGetArtifactIdMethod.invoke(artifact);
-                String version = (String) artifactGetVersionMethod.invoke(artifact);
+                String version = (String) artifactGetBaseVersionMethod.invoke(artifact);
                 String classifier = (String) artifactGetClassifierMethod.invoke(artifact);
 
                 if (library.getGroupId().equals(groupId) && library.getArtifactId().equals(artifactId))

--- a/core/src/main/java/com/alessiodp/libby/transitive/TransitiveDependencyHelper.java
+++ b/core/src/main/java/com/alessiodp/libby/transitive/TransitiveDependencyHelper.java
@@ -2,8 +2,10 @@ package com.alessiodp.libby.transitive;
 
 import com.alessiodp.libby.Library;
 import com.alessiodp.libby.LibraryManager;
+import com.alessiodp.libby.Util;
 import com.alessiodp.libby.classloader.IsolatedClassLoader;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 import java.io.IOException;
 import java.lang.reflect.Constructor;
@@ -14,6 +16,7 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Map.Entry;
 import java.util.Set;
 import java.util.stream.Stream;
 
@@ -46,7 +49,7 @@ public class TransitiveDependencyHelper {
     /**
      * Reflected getter methods of Artifact class
      */
-    private final Method artifactGetGroupIdMethod, artifactGetArtifactIdMethod, artifactGetBaseVersionMethod, artifactGetClassifierMethod;
+    private final Method artifactGetGroupIdMethod, artifactGetArtifactIdMethod, artifactGetVersionMethod, artifactGetBaseVersionMethod, artifactGetClassifierMethod;
 
     /**
      * LibraryManager instance, used in {@link #findTransitiveLibraries(Library)}
@@ -93,6 +96,8 @@ public class TransitiveDependencyHelper {
             artifactGetGroupIdMethod = artifactClass.getMethod("getGroupId");
             // org.eclipse.aether.artifact.Artifact#getArtifactId()
             artifactGetArtifactIdMethod = artifactClass.getMethod("getArtifactId");
+            // org.eclipse.aether.artifact.Artifact#getVersion()
+            artifactGetVersionMethod = artifactClass.getMethod("getVersion");
             // org.eclipse.aether.artifact.Artifact#getBaseVersion()
             artifactGetBaseVersionMethod = artifactClass.getMethod("getBaseVersion");
             // org.eclipse.aether.artifact.Artifact#getClassifier()
@@ -111,7 +116,7 @@ public class TransitiveDependencyHelper {
      * </p>
      * <p>
      * Note: The method merges the repositories from both the library manager and the given library
-     * for dependency resolution. And clones all relocations into transitive libraries.
+     * for dependency resolution. It also clones all relocations into transitive libraries.
      * </p>
      *
      * @param library The primary library for which transitive dependencies need to be found.
@@ -132,16 +137,20 @@ public class TransitiveDependencyHelper {
 
         Stream<String> repositories = Stream.of(globalRepositories, libraryRepositories).flatMap(Collection::stream);
         try {
-            Collection<?> artifacts = (Collection<?>) resolveTransitiveDependenciesMethod.invoke(transitiveDependencyCollectorObject,
+            Collection<?> resolvedArtifacts = (Collection<?>) resolveTransitiveDependenciesMethod.invoke(transitiveDependencyCollectorObject,
                 library.getGroupId(),
                 library.getArtifactId(),
                 library.getVersion(),
                 library.getClassifier(),
                 repositories);
-            for (Object artifact : artifacts) {
+            for (Object resolved : resolvedArtifacts) {
+                Entry<?, ?> resolvedEntry = (Entry<?, ?>) resolved;
+                Object artifact = resolvedEntry.getKey();
+                @Nullable String repository = (String) resolvedEntry.getValue();
+
                 String groupId = (String) artifactGetGroupIdMethod.invoke(artifact);
                 String artifactId = (String) artifactGetArtifactIdMethod.invoke(artifact);
-                String version = (String) artifactGetBaseVersionMethod.invoke(artifact);
+                String baseVersion = (String) artifactGetBaseVersionMethod.invoke(artifact);
                 String classifier = (String) artifactGetClassifierMethod.invoke(artifact);
 
                 if (library.getGroupId().equals(groupId) && library.getArtifactId().equals(artifactId))
@@ -153,7 +162,7 @@ public class TransitiveDependencyHelper {
                 Library.Builder libraryBuilder = Library.builder()
                                                         .groupId(groupId)
                                                         .artifactId(artifactId)
-                                                        .version(version)
+                                                        .version(baseVersion)
                                                         .isolatedLoad(library.isIsolatedLoad())
                                                         .loaderId(library.getLoaderId());
 
@@ -162,7 +171,30 @@ public class TransitiveDependencyHelper {
                 }
 
                 library.getRelocations().forEach(libraryBuilder::relocate);
-                library.getRepositories().forEach(libraryBuilder::repository);
+
+                if (repository != null) {
+                    // Construct direct download URL
+
+                    // Add ending "/" if missing
+                    if (!repository.endsWith("/")) {
+                        repository = repository + '/';
+                    }
+
+                    // TODO Uncomment the line below once LibraryManager#resolveLibrary stops resolving snapshots
+                    //      for every repository before trying direct URLs
+                    // Make sure the repository is added as fallback if the dependency isn't found at the constructed URL
+                    // libraryBuilder.repository(repository);
+
+                    // For snapshots, getVersion() returns version-timestamp-buildNumber instead of version-SNAPSHOT
+                    String version = (String) artifactGetVersionMethod.invoke(artifact);
+
+                    String partialPath = Util.craftPartialPath(artifactId, groupId, baseVersion);
+                    String path = Util.craftPath(partialPath, artifactId, version, classifier);
+
+                    libraryBuilder.url(repository + path);
+                } else {
+                    library.getRepositories().forEach(libraryBuilder::repository);
+                }
 
                 transitiveLibraries.add(libraryBuilder.build());
             }

--- a/core/src/test/java/com/alessiodp/libby/transitive/TransitiveDownloadingTest.java
+++ b/core/src/test/java/com/alessiodp/libby/transitive/TransitiveDownloadingTest.java
@@ -28,6 +28,15 @@ public class TransitiveDownloadingTest {
             .resolveTransitiveDependencies(true)
             .excludeTransitiveDependency(EXCLUDED_LIBRARY.getGroupId(), EXCLUDED_LIBRARY.getArtifactId())
             .build();
+    private static final Library BUNGEECORD = Library.builder()
+            .groupId("net{}md-5")
+            .artifactId("bungeecord-api")
+            .version("1.20-R0.2-SNAPSHOT")
+            .repository("https://oss.sonatype.org/content/repositories/snapshots")
+            .isolatedLoad(true)
+            .loaderId("bungeecord")
+            .resolveTransitiveDependencies(true)
+            .build();
 
     private LibraryManagerMock libraryManager;
 
@@ -47,6 +56,13 @@ public class TransitiveDownloadingTest {
         libraryManager.loadLibrary(MAVEN_RESOLVER_SUPPLIER);
 
         checkDownloadedDependencies();
+    }
+
+    @Test
+    public void snapshotLibraryTransitiveLoad() throws Exception {
+        libraryManager.loadLibrary(BUNGEECORD);
+
+        assertNotNull(libraryManager.getIsolatedClassLoaderById("bungeecord").loadClass("net.md_5.bungee.api.ProxyServer"));
     }
 
     @Test

